src/network/netdev/tuntap.c

```diff
@@ -137,6 +137,12 @@ static int netdev_create_tuntap(NetDev *netdev) {
         if (ioctl(fd, TUNSETIFF, &ifr) < 0)
                 return log_netdev_error_errno(netdev, errno, "TUNSETIFF failed: %m");
 
+        if (t->multi_queue) {
```

Comment:
here you check for `multi_queue`, not for keep_fd?

Reply:
Yes, it's intentional: networkd never reads packets from its devices, so it would be more correct to always disable the queue right after device creation / attaching to existing device to prevent packet dropping even for the short time of device configuration. So I would even remove `if (t->multi_queue)` and do it always, but kernel doesn't allow to disable queue on non-multi queue devices. It would be best to attach in disabled queue mode, but kernel doesn't have one, so the best we can do is disable the queue right after attaching.

Reply:
hmm, so this requires an in-code comment.

But what I am not grokking: this detaches the queue, but if a carrier comes, then something needs to attach the queue again? what am i missing here?

(my insight into tun/tap behaviour is a bit superficial, in case you didn't notice)

Reply:
When we open tun/tap device, kernel creates a queue for each open FD.

Link is considered to have carrier if its open FD count is not zero.

When packet arrives, kernel selects a queue to send the packet to.  From a quick look at the kernel sources the logic is something like "take current CPU ID and map it to index in active queue list". So if some packets are handled by CPU #N and N maps to our queue, all these packets will be sent to our queue - even if no one reads it.

So from my view the whole picture should be the following:
* I create a tun/tap device using networkd, which properly configures it and leaves FD open, but disables its queue. So kernel considering the device as connected.
* Then I start some rootless service which opens the device and kernels starts send all packets to its queue(s).
* When the service stops, we have one open FD (by networkd) and 0 active and 1 disabled queue (networkd's). Kernel remains to consider link as active, and continues to route packets to it, but they are effectively dropped since we don't have active queues now. And this is desired behaviour for us - for example in case of VPN. In other case we might need to create some blackhole routing table with lower priority to make kernel send packets to it when the main interface lose its carrier.
* When the service starts again, we have one active queue and packets start flow to it.